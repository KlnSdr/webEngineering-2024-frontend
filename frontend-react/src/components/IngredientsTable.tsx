import React from "react";

/**
 * Interface representing a single ingredient.
 * @typedef {Object} Ingredient
 * @property {string} name - The name of the ingredient.
 * @property {number} amount - The amount of the ingredient.
 * @property {string} unit - The unit of the ingredient amount.
 */
interface Ingredient {
  name: string;
  amount: number;
  unit: string;
}

/**
 * Props for the IngredientsTable component.
 * @typedef {Object} IngredientsTableProps
 * @property {Ingredient[]} ingredients - The list of ingredients to display in the table.
 */
interface IngredientsTableProps {
  ingredients: Ingredient[];
}

/**
 * IngredientsTable component renders a table of ingredients.
 * @param {IngredientsTableProps} props - The props for the component.
 * @returns {JSX.Element} The rendered IngredientsTable component.
 */
const IngredientsTable: React.FC<IngredientsTableProps> = ({ ingredients }) => {
  return (
    <table>
      <thead>
        <tr>
          <th>Zutat</th>
          <th>Menge</th>
        </tr>
      </thead>
      <tbody>
        {ingredients.map((ingredient, index) => (
          <tr key={index}>
            <td>{ingredient.name}</td>
            <td>{ingredient.amount} {ingredient.unit}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default IngredientsTable;