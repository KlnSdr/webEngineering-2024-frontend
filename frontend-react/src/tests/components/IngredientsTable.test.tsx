import React from "react";
import {render, screen, within} from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import IngredientsTable from "../../components/IngredientsTable";

describe("IngredientsTable Component", () => {
    const ingredients = [
        {
            name: "Test Ingredient",
            amount: 1,
            unit: "kg"
        }
    ];

    test("renders the correct ingredients", () => {
        const {asFragment} = render(<IngredientsTable ingredients={ingredients} />);

        const tableElement = screen.getByRole("table");
        expect(tableElement).toBeInTheDocument();

        const headRowElement = screen.getAllByRole("row")[0];
        expect(headRowElement).toBeInTheDocument();

        const headDataElements = within(headRowElement).getAllByRole("columnheader");
        expect(headDataElements).toHaveLength(2);
        expect(headDataElements[0]).toHaveTextContent("Zutat");
        expect(headDataElements[1]).toHaveTextContent("Menge");

        const bodyRowElement = screen.getAllByRole("row")[1];
        expect(bodyRowElement).toBeInTheDocument();

        const bodyDataElements = within(bodyRowElement).getAllByRole("cell");
        expect(bodyDataElements).toHaveLength(2);
        expect(bodyDataElements[0]).toHaveTextContent("Test Ingredient");
        expect(bodyDataElements[1]).toHaveTextContent("1 kg");
        
        expect(asFragment()).toMatchSnapshot();
    });
});
