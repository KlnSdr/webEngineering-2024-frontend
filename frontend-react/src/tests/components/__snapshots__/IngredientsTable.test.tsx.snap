// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IngredientsTable Component renders the correct ingredients 1`] = `
<DocumentFragment>
  <table>
    <thead>
      <tr>
        <th>
          Zutat
        </th>
        <th>
          Menge
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Test Ingredient
        </td>
        <td>
          1 kg
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
